### PR TITLE
ET-4775 check filter against property schema

### PR DESCRIPTION
**Pull request by janpetschexain**
**Reference**

- [ET-4775]

**Summary**

- impl filter validation wrt the indexed properties schema
- improve filter serde tests & add filter validation tests

[ET-4775]: https://xainag.atlassian.net/browse/ET-4775?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: web-api/src/personalization/filter.rs [hunk 1]
```diff
@@ -444,227 +519,114 @@ mod tests {
...
             ("$lt", CompareOp::Lt),
             ("$lte", CompareOp::Lte),
         ] {
             for number in [json!(42_u64), json!(-42_i64), json!(42_f64)] {
                 assert_eq!(
-                    serde_json::from_str::<CompareWith>(
-                        &json!({ operation.0: number }).to_string()
-                    )
-                    .unwrap(),
+                    serde_json::from_str::<CompareWith>(&format!(r#"{{ "{op}": {number} }}"#))
```
Comment (rustonaut) [inline web-api/src/personalization/filter.rs:586]: :wrench:  Why not just use `json`? 

(here and below where you sometimes explicitly changed away from it).

**Review by rustonaut — COMMENTED** [review]
looks mostly good to me

the only thing is we check `(operator, value)` and `(value, property_type)` disjoint with implicit assumptions about the other. 

This is fine for now as it works and changing it is a major rework.

But in the end which `(value, property_type)` combinations are valid is dependent on `operator`.

File: web-api/src/personalization/filter.rs (hunk 1 above)
Comment (janpetschexain) [reply web-api/src/personalization/filter.rs:586]: i'm a bit undecided about this, in the end the outcome is of course the same if we use `format!()` or `json!().to_string()`. currently i'm leaning more towards the former as this seems a bit closer to what we would get in a request written by a user. but i also the that the additional escaping in the formatting is not nice. should we go back to the latter, wdyt?

Comment (rustonaut) [reply web-api/src/personalization/filter.rs:586]: I think  `json!` is more readable and guarantees that the json itself is always well formed and as such should be preferred in this context as we are not testing the json parsing itself.

Any performance differences are in this use-case irrelevant AFIK.

**Comment by janpetschexain** [comment]
> looks mostly good to me
> 
> the only thing is we check `(operator, value)` and `(value, property_type)` disjoint with implicit assumptions about the other.
> 
> This is fine for now as it works and changing it is a major rework.
> 
> But in the end which `(value, property_type)` combinations are valid is dependent on `operator`.

yes, in the end it is the same line of thought like that we do additional validation on a property with `schema.validaty_property()`, just applied to filters this time. the `schema.validate_filter()` is of course not enough/complete on its own.

**Review by rustonaut — APPROVED** [review]
lgtm

**Pull request merged**